### DISCUSSION TITLE
cups use cases with IPPeve

### DISCUSSION THREAD

**Discussion by robertmx** [OpenPrinting/cups#635] [General] [answered]
Dear developers,

I am wondering what use cases does cups solve with cups moving towards only talking to IPPeve-capable printers or drivers? Wouldn't it be simpler to just send the print jobs directly to the printer instead of talking via IPPeve to cups, which in turn talks to the printer using the same protocol?
To be clear: I don't want to be negative or dress down the project I have been using for years as a print server for legacy printers. I am just curious about the advantages of still using cups instead of directly talking IPPeve?

Cheers

**Comment by michaelrsweet** [comment, marked as answer]
@robertmx TL;DR: So there are reasons for this historically but for CUPS 3.0 we are actually looking at using D-Bus for local printing on Linux.

Fundamentally CUPS needs to do the following:

- Spool print jobs (so you can queue up multiple things to print as needed and not have apps waiting or competing for access to printers)
- Convert jobs to a format the printer can understand (not app printers support PDF and JPEG), and apply common options locally as needed - copies and page-ranges for raster printers, number-up for printers that don't support it natively, etc.
- Report printer issues (out of paper/ink/toner, paper jam, etc.) and handle authentication/authorization
- Manage the print queues and print jobs (pause/resume a printer, cancel a job, etc.)

Using IPP between the application and spooler is convenient because it minimizes the number of conversions needed and you can print to a local or remote print service with the same code. But as you note this isn't *required* just an implementation detail. For CUPS 3.0 we are looking at adding support for D-Bus (on Linux/*BSD) and XPC (macOS/iOS) as transports so that we can more easily fit in the "modern" security container architectures of these platforms and (more importantly) talk to a local print server running in the user's session so that per-user security credentials and other interactions happen in the user's session and not in a cupsd process running as root.

If you look at the [slides](https://ftp.pwg.org/pub/pwg/liaison/openprinting/presentations/cups-plenary-may-2022.pdf) from the last OpenPrinting Summit you will see where CUPS 3.0 is going.

**Reply by robertmx** [reply]
Thank you very much for you comprehensive and concise answer, outlining the various use cases. That will come in handy when planning the next CUPS deployment.

Cheers
